caffeine/src/main/java/com/github/benmanes/caffeine/cache/LocalCacheFactory.java

```diff
@@ -80,14 +84,27 @@ static String getClassName(Caffeine<?, ?> builder) {
 
   static <K, V> BoundedLocalCache<K, V> loadFactory(Caffeine<K, V> builder,
       @Nullable AsyncCacheLoader<? super K, V> cacheLoader, boolean async, String className) {
+    return CONSTRUCTORS.computeIfAbsent(className, LocalCacheFactory::newConstructor)
+        .construct(builder, cacheLoader, async);
```

Comment:
A `computeIfAbsent` can become a [bottleneck](https://github.com/ben-manes/caffeine/wiki/Benchmarks#compute) if present due to pessimistic locking. Our feedback led to some partial improvements in Java 9, but it still remains if unlucky. Caffeine will always do a full prescreen by being optimistic that the entry is present, which would be my preference for CHM to adopt. You can read Doug's [rational](https://markmail.org/message/dixewqcvbbs5gzi2) wrt safepoints, bias locking, etc.

In these cases you can avoid the problem by a lock-free prescreen like,

```java
var factory = CONSTRUCTORS.get(className);
if (factory == null) {
  factory = CONSTRUCTORS.computeIfAbsent(className, LocalCacheFactory::newConstructor);
}
return factory.construct(builder, cacheLoader, async);
```